actors/builtin/market/market_actor.go

```diff
@@ -289,65 +289,118 @@ func (A Actor) VerifyDealsForActivation(rt Runtime, params *VerifyDealsForActiva
+				continue
 			}
 
-			proposal, err := getDealProposal(msm.dealProposals, dealID)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to get dealId %d", dealID)
+			success := true
+			for _, dealID := range req.DealIDs {
+				// This construction could be replaced with a single "update deal state" state method, possibly batched
+				// over all deal ids at once.
+				_, found, err := msm.dealStates.Get(dealID)
+				if err != nil {
+					rt.Log(vmr.ERROR, "failed to get deal state for activation, dealId: %d, sector number: %v, "+
+						"err: %w", dealID, req.SectorNumber, err)
+					success = false
+					break
+				}
 
-			propc, err := proposal.Cid()
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to calculate proposal CID")
+				if found {
+					rt.Log(vmr.ERROR, "deal %d for sector %v already included in another sector", dealID, req.SectorNumber)
+					success = false
+					break
+				}
 
-			has, err := msm.pendingDeals.Get(adt.CidKey(propc), nil)
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to get pending proposal %v", propc)
+				proposal, err := getDealProposal(msm.dealProposals, dealID)
+				if err != nil {
+					rt.Log(vmr.ERROR, "failed to get deal proposal for activation, dealId: %d, sector number: %v, err: %w", dealID,
+						req.SectorNumber, err)
+					success = false
+					break
+				}
 
-			if !has {
-				rt.Abortf(exitcode.ErrIllegalState, "tried to activate deal that was not in the pending set (%s)", propc)
+				propc, err := proposal.Cid()
+				if err != nil {
+					rt.Log(vmr.ERROR, "failed to calculate proposal CID for activation, dealId: %d, sector number: %v, err: %w", dealID,
+						req.SectorNumber, err)
+					success = false
+					break
+				}
+
+				has, err := msm.pendingDeals.Get(adt.CidKey(propc), nil)
+				if err != nil {
+					rt.Log(vmr.ERROR, "failed to get pending proposal for activation, dealId: %d, sector number: %v, err: %w", dealID,
+						req.SectorNumber, err)
+					success = false
+					break
+				}
+
+				if !has {
+					rt.Log(vmr.ERROR, "tried to activate deal that was not in pending state, dealId: %d, sector number: %v, err: %w", dealID,
+						req.SectorNumber, err)
+					success = false
+					break
+				}
 			}
 
-			err = msm.dealStates.Set(dealID, &DealState{
-				SectorStartEpoch: currEpoch,
-				LastUpdatedEpoch: epochUndefined,
-				SlashEpoch:       epochUndefined,
-			})
-			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to set deal state %d", dealID)
+			if success {
+				for _, dealID := range req.DealIDs {
+					err := msm.dealStates.Set(dealID, &DealState{
+						SectorStartEpoch: currEpoch,
+						LastUpdatedEpoch: epochUndefined,
+						SlashEpoch:       epochUndefined,
+					})
+
+					builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to set deal state for activation, dealId: %d, sector number: %v, err: %w",
```

Comment:
@anorth @ZenGround0 

Unfortunately, I think we need to abort here for atomicity.

If we don't do this, we could end up in a situation where some deals are set and some others fail which would leave the sector "fragmented".

Reply:
I'm in support of propagating this illegal state error to ConfirmSectorProofsValid and failing there instead of just logging.  If our prove commits expose illegal market state there is a big problem that should be brought to light rather than patched over in the interest of saving the miner some funds.